drivers: digital-io: max22200 : Add driver support for MAX22200

RaduSabau1 opened the pull request:
## Pull Request Description

Add driver support for MAX22200.

The MAX22200 is an octal 36V serial-controlled solenoid driver.
Each channel features a low impedance (200mΩ typ) push-pull output stage with sink-and-source driving
capability and up to 1ARMS driving current. A serial interface (SPI) that also supports daisy-chain configurations is
provided to individually control each channel

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (ret)
+		return ret;
+
+	if (one_byte) {
```
nit: can we do here an if else and return only once? In this manner we avoid duplicated `return 0`

RaduSabau1 replied:
Sure.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+		.bytes_number = 1,
+		.cs_change = 1,
+	};
+	int ret;
```
nit: i'd put this declaration first, before the definition of struct.

RaduSabau1 replied:
Moving it.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <errno.h>
```
no_os_error.h

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+		.bytes_number = 1,
+		.cs_change = 1,
+	};
+	int ret;
```
ditto.

RaduSabau1 replied:
ofc.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (ret)
+		return ret;
+
+	if (!one_byte) {
```
either use brackets for both if and else statement, either you don't for both.

RaduSabau1 replied:
Adding brackets to the else statement.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+
+	reg_val &= ~mask;
+	reg_val |= mask & val;
+	return max22200_reg_write(desc, reg, reg_val, one_byte);
```
nit: i'd add extra line before last return of the function.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+int max22200_set_dev_freq(struct max22200_desc *desc,
+			  enum max22200_device_freq device_freq)
+{
+	if (device_freq > 1)
```
since the `device_freq` is of type `enum max22200_device_freq` I don't think this check is useful.

RaduSabau1 replied:
removing it.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+		return ret;
+
+	return max22200_reg_update(desc, MAX22200_CFG_CH(ch),
+				   MAX22200_HIT_T_MASK,
```
indentation is pretty odd. this applies to multiple returns throughout the code.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (ch > MAX22200_CHANNELS - 1)
+		return -EINVAL;
+
+	if (trig > 1)
```
ditto. redundant check.

RaduSabau1 replied:
Removing it.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (ch > MAX22200_CHANNELS - 1)
+		return -EINVAL;
+
+	if (ch_freq > 3)
```
same. ch_freq has a specific type.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (ch > MAX22200_CHANNELS - 1)
+		return -EINVAL;
+
+	if (ch_enable > 3)
```
ditto.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (ret)
+		return ret;
+
+	*enabled =  no_os_field_get(MAX22200_CH_ENABLE_MASK(ch_enable),
```
nit: extra space.

RaduSabau1 replied:
removing it.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+			goto fault_err;
+	}
+
+	no_os_udelay(2500);
```
meaning of this delay?

RaduSabau1 replied:
Adding a comment specifying why this delay is needed.

rbolboac commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+		ret = no_os_gpio_direction_output(descriptor->enable_desc,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			goto enable_err;
```
cmd_err, enable_desc is not allocated in this case

rbolboac commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+		ret = no_os_gpio_direction_output(descriptor->trig_desc,
+						  NO_OS_GPIO_LOW);
+		if (ret)
+			goto trig_err;
```
enable_err, trig_desc is not allocated in this case

RaduSabau1 replied:
the if condtion above checks if trig_desc is initialized there by allocated, same for the fault_err handling.

rbolboac replied:
oops, my bad. I misread the code.

rbolboac commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (descriptor->fault_desc) {
+		ret = no_os_gpio_direction_input(descriptor->fault_desc);
+		if (ret)
+			goto fault_err;
```
trig_err, fault_desc is not allocated in this case

rbolboac commented on drivers/digital-io/max22200/iio_max22200.h:
```diff
@@ -0,0 +1,130 @@
+/** Free resources allocated by the init function. */
+int max22200_iio_remove(struct max22200_iio_desc *);
+
+#endif /** IIO_MAX22200_H */
```
missing new line at the end of file

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1234 @@
+
+	for (i = 0; i < MAX22200_CHANNELS; i++)
+		if (ch_enable[i])
+			enabled_ch++;;;
```
remove extra 2 semi-columns

rbolboac commented on drivers/digital-io/max22200/iio_max22200.h:
```diff
@@ -0,0 +1,130 @@
+#include <stdint.h>
+#include <stdbool.h>
+#include "iio.h"
+#include "iio_types.h"
```
not needed

RaduSabau1 replied:
removing it.

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1234 @@
+
+static struct iio_attribute max22200_global_attrs[] = {
+	{
+		.name = "device_frequency",
```
is this the sampling frequency of the device? let's try and use some generic attribute names if they apply. check this list: https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-bus-iio

RaduSabau1 replied:
it's the chopping frequency of the device, can switch name to chopping_frequency if needed.

rbolboac replied:
yes, the name would be cleared

RaduSabau1 replied:
Changing to chopping_frequency then.

RaduSabau1 commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,958 @@
+	if (ret)
+		return ret;
+
+	*(desc->ch_config + (ch / 2)) = ch_op_mode;
```
Channels are configured 2 by 2, will add a macro for it.

CiprianRegus replied:
`desc->ch_config[ch / 2] = ch_op_mode`

CiprianRegus replied:
Also, a comment would help more. Using the macro doesn't make this easier to read.

RaduSabau1 commented:
V2

- Fixed includes.
- Fixed typos and coding style.
- Removed unnecessary checks.
- Addressed other comments.

rbolboac commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,939 @@
+	if (hit_current > MAX22200_HIT_MAX_VAL)
+		return -EINVAL;
+
+	ret = max22200_reg_update(desc, MAX22200_CFG_CH(ch),
+				  MAX22200_HIT_MASK,
+				  no_os_field_prep(MAX22200_HIT_MASK,
+						  (uint32_t)hit_current), false);
```
I would let astyle formatting take care of this. basically just write everything on 1 line and then use the VSC for formatting... see everywhere where it applies. it should look something like this:
```
	ret = max22200_reg_update(desc, MAX22200_CFG_CH(ch), MAX22200_HIT_MASK,
				  no_os_field_prep(MAX22200_HIT_MASK, (uint32_t)hit_current), false);
```

rbolboac commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,939 @@
+	int ret;
+	uint32_t reg_val;
+
+	if (fault_mask < 17 || fault_mask > 23)
+		return -EINVAL;
```
can be removed

RaduSabau1 replied:
Forgot about it the last time.

RaduSabau1 commented:
V3

- Fixed coding style.
- Changed name of attribute from device frequency to chopping frequency.
- Addressed other comments.

RaduSabau1 commented:
V4

- Rebased with master.

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,922 @@
+ * 		     false - Whole 32bit register will be returned through val.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max22200_reg_read(struct max22200_desc *desc, uint32_t reg, uint32_t *val,
```
having a second look, do we want to deviate from the "standard" read/write function prototypes by adding the `one_byte` parameter? Having a look at the datasheet, all registers are 32bits wide. Can't we just propely adjust the 32-bit register values prior to calling the read/write functions?

RaduSabau1 replied:
Well with this chip you can either read/write the whole 32-bit register or only the LSB, and since the chip provides this functionality I thought the driver should as well.
There is also a possibility to use a bit in the reg value which can be used instead of the one_byte parameter, any thoughts about this?

amiclaus commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,922 @@
+	for (i = 0; i < MAX22200_CHANNELS_CONFIG; i++)
+		status_reg |= no_os_field_prep(MAX22200_CH_MODE_MASK(i),
+					       *(init_param->ch_config + i));
+
```
nit: extra line

amiclaus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1234 @@
+
+	switch (priv) {
+	case MAX22200_IIO_HIT_CURRENT:
+		if (ret)
```
this is odd. move it outside the switch case.

RaduSabau1 commented:
V5

- Removed one_byte parameter in read/write functions.

amiclaus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1230 @@
+			}
+		}
+
+		if (i == NO_OS_ARRAY_SIZE(max22200_ch_drive_avail))
```
all these checks can be dropped if you return directly in the for loop. same applies in other functions of the driver. This is just a suggestion.

RaduSabau1 replied:
commited.

rbolboac commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,910 @@
+int max22200_set_ch_state(struct max22200_desc *desc, uint32_t ch,
+			  enum max22200_ch_state ch_state)
+{
+	if (ch > MAX22200_CHANNELS - 1)
```
I would rename this to MAX22200_MAX_CHN_IDX and give it a value of 7, and make the updates everywhere.

RaduSabau1 replied:
will do.

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1212 @@
+	if (val > 1)
+		val = 1;
+
+	return max22200_set_trig_state(desc->max22200_desc, val);
```
so here the function takes an enumeration as parameter and you are giving a uint32_t value. I would replace this code with something like:
```
return max22200_set_trig_state(desc->max22200_desc, val ? MAX22200_TRIG_ON : MAX22200_TRIG_OFF);
```
which shows better what is the intended purpose and what val represents
see in all APIs where applicable

RaduSabau1 replied:
sure.

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1212 @@
+	if (val > 1)
+		val = 1;
+
+	return max22200_fault_mask_set(desc->max22200_desc, priv, val);
```
return max22200_fault_mask_set(desc->max22200_desc, priv, val ? true : false);

RaduSabau1 commented:
V6

- Removed API call related warnings.
- Addressed other comments.

buha commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,910 @@
+/**
+ * @brief Read data from desired register for MAX22200
+ * @param desc - MAX22200 device descriptor.
+ * @param reg - Register value to read data from.
```
Rephrase to register number or register address

buha commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,910 @@
+ * @brief Read data from desired register for MAX22200
+ * @param desc - MAX22200 device descriptor.
+ * @param reg - Register value to read data from.
+ * @param val - Value read from the register.
```
something is odd about this phrasing

buha commented on drivers/digital-io/max22200/iio_max22200.h:
```diff
@@ -0,0 +1,131 @@
+};
+
+static const char *const max22200_scale_avail[2] = {
+	"FULLSCALE",
```
I'm not sure using caps is a good idea, i think i haven't come across of iio device drivers that use caps in attribute fields

RaduSabau1 commented on drivers/digital-io/max22200/max22200.c:
```diff
@@ -0,0 +1,910 @@
+		xfer.bytes_number = MAX22200_FRAME_SIZE;
+		no_os_put_unaligned_le32(val, desc->buff);
+	} else {
+		desc->buff[0] = val;
```
there is an if condition above that checks if val is greater than the max value for uint8_t in case of just one byte write operation.

buha replied:
I mean, I briefly checked the datasheet register map and all registers are 4 byte long. What's the catch ? Can you do a partial write or read of a register?

RaduSabau1 replied:
Table 9 in max22200 datasheet explains the situation.

buha commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1221 @@
+	},
+	{
+		.name = "scale_available",
+		.show = max22200_iio_read_scale_available,
```
You have a function for each attribute's show/store. This works but it can be written in a more compact way.

For example, with regards to these `...available` attributes, they all return a string directly, so you could have 1 single show function that handles all available attributes and differentiates them using an ID that you pass through `iio_attribute.priv`.

So instead of having 20 functions that do more or less the same think, you end up with 3/4/5, best case 1/2 for simple devices.

RaduSabau1 replied:
Tried to group as much as possible. I couldn't add ch_mode_available with the others since it needs its own separate priv.

RaduSabau1 commented:
V7

- Grouped attributes to be showed by the same API using different priv for each one of them.
- Using bool instead of enum for on/off states.
- Addressed other comments.

RaduSabau1 commented:
V8

- Fixed name of API for chopping frequency.
- Fixed build error.

RaduSabau1 commented:
V9

- Added iio_write_raw API.
- Using uint32_t variables instead of uint8_t for hit and hold attribute.
- Addressed comments related to unnecessary checks.

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1090 @@
+	if (ret)
+		return ret;
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1,
```
casting from bool to int32_t is not a good idea. please make the update also for this case

RaduSabau1 replied:
Using different approach then.

rbolboac reviewed (approved):
after fixing that, can be merged from my point of view

CiprianRegus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1090 @@
+{
+	struct max22200_iio_desc *desc = dev;
+	enum max22200_trig ch_trig;
+	uint32_t val;
```
Declare this variable as `int32_t` so you avoid casting

RaduSabau1 commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1090 @@
+	int ret;
+
+	ret = max22200_get_ch_hold(desc->max22200_desc, channel->address,
+				   (uint8_t)&val);
```
Fixing it.

RaduSabau1 commented:
V10

- Fixed wrong casting.
- Addressed other comments.

CiprianRegus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1091 @@
+	switch (priv) {
+	case MAX22200_IIO_HIT_CURRENT:
+		return iio_format_value(buf, len, IIO_VAL_INT, 1,
+					(int32_t *)&hit_c);
```
Casting to a pointer of a wider type is wrong. It will result in accessing memory beyond where `hit_c` is stored.

RaduSabau1 replied:
My bad, I misunderstood the earlier comment on this same API, fixing it.

CiprianRegus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1091 @@
+	if (ret)
+		return ret;
+
+	val = (int32_t)ch_trig;
```
no need to cast

CiprianRegus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1091 @@
+	if (ret)
+		return ret;
+
+	val = (int32_t)chopping_freq;
```
useless cast

CiprianRegus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1091 @@
+				  intptr_t priv)
+{
+	struct max22200_iio_desc *desc = dev;
+	uint32_t val;
```
`int32_t`

RaduSabau1 replied:
Fixing and applying where it fits.

CiprianRegus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1091 @@
+		intptr_t priv)
+{
+	struct max22200_iio_desc *desc = dev;
+	uint32_t val;
```
`int32_t`

amiclaus reviewed (approved):
a changelog would be nice after your latest push.

RaduSabau1 commented:
V11

- Fixed bad memory access.
- Addressed other comments.

CiprianRegus commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1091 @@
+	int ret;
+
+	ret = max22200_get_ch_hit(desc->max22200_desc, channel->address,
+				  (uint8_t *)&hit_c, (uint8_t *)&hit_t);
```
This will lead to different results based on the cpu's endianess. Please stop casting pointers to different widths, it's not safe to do so, and there is no point in doing it.

RaduSabau1 replied:
Won't cast then since there is no point in doing it.

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1091 @@
+	uint32_t hit_c, hit_t;
+	int ret;
+
+	ret = max22200_get_ch_hit(desc->max22200_desc, channel->address,
```
get_ch_hit expect uin8_t* and you are giving uint32_t*. please use the same type of pointer. see everywhere where applicable

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1098 @@
+		intptr_t priv)
+{
+	struct max22200_iio_desc *desc = dev;
+	uint32_t val;
```
can be int32_t and cast can be removed

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1098 @@
+				 intptr_t priv)
+{
+	struct max22200_iio_desc *desc = dev;
+	uint32_t hit_c, hit_t;
```
can be int32_t and cast can be removed

rbolboac commented on drivers/digital-io/max22200/iio_max22200.c:
```diff
@@ -0,0 +1,1098 @@
+				  intptr_t priv)
+{
+	struct max22200_iio_desc *desc = dev;
+	uint32_t val;
```
can be int32_t and cast can be removed

RaduSabau1 commented:
V13

- Using variable casting instead of pointer casting.
